client/components/site-selector/index.jsx

```diff
@@ -62,50 +64,160 @@ const SiteSelector = React.createClass( {
 
 	getInitialState() {
 		return {
-			search: ''
+			search: '',
+			highlightedIndex: -1,
+			showSearch: false
 		};
 	},
 
+	reset() {
+		if ( this.state.search && this.refs.siteSearch ) {
+			this.refs.siteSearch.clear();
+		} else {
+			this.setState( this.getInitialState() );
+		}
+	},
+
 	onSearch( terms ) {
-		this.setState( { search: terms } );
+		this.setState( {
+			search: terms,
+			highlightedIndex: ( terms ? 0 : -1 ),
+			showSearch: ( terms ? true : this.state.showSearch ),
+			isKeyboardEngaged: false,
+		} );
+	},
+
+	componentDidUpdate( prevProps, prevState ) {
+		if ( this.state.isKeyboardEngaged && prevState.highlightedIndex !== this.state.highlightedIndex ) {
+			this.scrollToHightlightedSite();
+		}
+	},
+
+	scrollToHightlightedSite() {
+		const selectorElement = ReactDom.findDOMNode( this.refs.selector );
+		if ( selectorElement ) {
+			const highlightedSiteElement = ReactDom.findDOMNode( this.refs.highlightedSite );
+			if ( highlightedSiteElement ) {
+				scrollIntoView( highlightedSiteElement, selectorElement, {
+					onlyScrollIfNeeded: true
+				} );
+			} else {
+				selectorElement.scrollTop = 0;
+			}
+		}
 	},
 
-	onSiteSelect( siteSlug, event ) {
-		this.closeSelector();
+	onKeyDown( event ) {
```

Comment:
I extracted some logic to `computeHighlightedSite`. Looks better now for me, what do you think?